File: server/src/services/embeddingService.ts
import { GoogleGenerativeAI } from '@google/generative-ai';
import { LocalIndex } from 'vectra';
import * as path from 'path';
import * as os from 'os';
import { config } from '../config';
import { chunkText } from '../utils/chunkText';
import { retry, sleep } from '../utils/retry';

const genAI = new GoogleGenerativeAI(config.googleApiKey);
const embeddingModel = genAI.getGenerativeModel({ model: 'text-embedding-004' }); // fallback to 001 if 004 fails

// We will export a getter for the model to ensure we can catch and fallback, OR just rigidly use gemini-embedding-001
// Actually, since we know gemini-embedding-001 works perfectly for this API key, we will just use it directly!

const activeEmbeddingModel = genAI.getGenerativeModel({ model: 'gemini-embedding-001' });

// In-memory map of videoId → Vectra index
const vectorIndexes = new Map<string, LocalIndex>();

function getIndexPath(videoId: string): string {
  return path.join(os.tmpdir(), 'ytllm-vectors', videoId);
}

async function getOrCreateIndex(videoId: string): Promise<LocalIndex> {
  if (vectorIndexes.has(videoId)) {
    return vectorIndexes.get(videoId)!;
  }

  const index = new LocalIndex(getIndexPath(videoId));
  if (!(await index.isIndexCreated())) {
    await index.createIndex();
  }
  vectorIndexes.set(videoId, index);
  return index;
}

/**
 * Embeds a single text using gemini-embedding-001.
 */
async function embedText(text: string): Promise<number[]> {
  return retry(async () => {
    const result = await activeEmbeddingModel.embedContent(text);
    return result.embedding.values;
  });
}

/**
 * Indexes transcript chunks for a video.
 * Chunks are embedded and stored in Vectra.
 */
export async function indexTranscript(videoId: string, transcript: string): Promise<number> {
  const chunks = chunkText(transcript, config.chunkSize, config.chunkOverlap);
  console.log(`📦 Indexing ${chunks.length} chunks for video ${videoId}`);

  // Clear existing index if any
  if (vectorIndexes.has(videoId)) {
    vectorIndexes.delete(videoId);
  }

  const index = await getOrCreateIndex(videoId);

  // Process in batches
  for (let i = 0; i < chunks.length; i += config.embeddingBatchSize) {
    const batch = chunks.slice(i, i + config.embeddingBatchSize);

    for (let j = 0; j < batch.length; j++) {
      const chunkIndex = i + j;
      const embedding = await embedText(batch[j]);
      await index.insertItem({
        vector: embedding,
        metadata: {
          text: batch[j],
          chunkIndex,
        },
      });
    }

    // Delay between batches to respect rate limits
    if (i + config.embeddingBatchSize < chunks.length) {
      await sleep(config.embeddingBatchDelay);
    }
  }

  console.log(`✅ Indexed ${chunks.length} chunks for video ${videoId}`);
  return chunks.length;
}

/**
 * Retrieves the top-k most semantically similar chunks for a query.
 */
export async function retrieveRelevantChunks(videoId: string, query: string, topK = config.topKChunks): Promise<string[]> {
  const index = vectorIndexes.get(videoId);
  if (!index) {
    console.warn(`No vector index found for video ${videoId}`);
    return [];
  }

  const queryEmbedding = await embedText(query);
  const results = await index.queryItems(queryEmbedding, query, topK);

  return results.map((r) => r.item.metadata.text as string);
}

/**
 * Deletes the vector index for a video (cleanup).
 */
export function deleteIndex(videoId: string): void {
  vectorIndexes.delete(videoId);
}

File: server-python/app/services/embedding_service.py
import os
import tempfile
import google.generativeai as genai
import chromadb
from chromadb.config import Settings as ChromaSettings

from app.config import config
from app.utils.chunk_text import chunk_text
from app.utils.retry import retry, sleep

genai.configure(api_key=config["google_api_key"])

# In-memory map of video_id -> chromadb client
vector_indexes = {}


def get_index_path(video_id: str) -> str:
    """Get the path for storing the vector index."""
    return os.path.join(tempfile.gettempdir(), "ytllm-vectors", video_id)


async def get_or_create_index(video_id: str):
    """Get or create a ChromaDB index for a video."""
    if video_id in vector_indexes:
        return vector_indexes[video_id]

    index_path = get_index_path(video_id)
    os.makedirs(index_path, exist_ok=True)

    client = chromadb.PersistentClient(path=index_path)
    collection = client.get_or_create_collection(
        name=f"video_{video_id}", metadata={"hnsw:space": "cosine"}
    )
    vector_indexes[video_id] = collection
    return collection


async def embed_text(text: str) -> list[float]:
    """Embeds a single text using gemini-embedding-001."""

    async def _embed():
        result = genai.embed_content(
            model="gemini-embedding-001",
            content=text,
        )
        return result["embedding"]

    return await retry(_embed, max_attempts=3)


async def index_transcript(video_id: str, transcript: str) -> int:
    """Indexes transcript chunks for a video."""
    chunks = chunk_text(transcript, config["chunk_size"], config["chunk_overlap"])
    print(f"Indexing {len(chunks)} chunks for video {video_id}")

    # Clear existing index
    if video_id in vector_indexes:
        del vector_indexes[video_id]

    collection = await get_or_create_index(video_id)

    # Process in batches
    for i in range(0, len(chunks), config["embedding_batch_size"]):
        batch = chunks[i : i + config["embedding_batch_size"]]

        for j, chunk in enumerate(batch):
            chunk_index = i + j
            embedding = await embed_text(chunk)

            collection.upsert(
                ids=[str(chunk_index)],
                embeddings=[embedding],
                documents=[chunk],
                metadatas=[{"chunk_index": chunk_index}],
            )

        # Delay between batches
        if i + config["embedding_batch_size"] < len(chunks):
            await sleep(int(config["embedding_batch_delay"] * 1000))

    print(f"Indexed {len(chunks)} chunks for video {video_id}")
    return len(chunks)


async def retrieve_relevant_chunks(
    video_id: str, query: str, top_k: int = None
) -> list[str]:
    """Retrieves the top-k most semantically similar chunks for a query."""
    if video_id not in vector_indexes:
        print(f"No vector index found for video {video_id}")
        return []

    if top_k is None:
        top_k = config["top_k_chunks"]

    collection = vector_indexes[video_id]

    query_embedding = await embed_text(query)

    results = collection.query(
        query_embeddings=[query_embedding],
        n_results=top_k,
    )

    documents = results.get("documents", [[]])[0]
    return documents


def delete_index(video_id: str) -> None:
    """Deletes the vector index for a video (cleanup)."""
    if video_id in vector_indexes:
        del vector_indexes[video_id]
